sdk/eventhub/azure-eventhubs-aio-partitionmanager-blobstorage/azure/eventhub/aio/partitionmanager/blobstorage/blobstoragepm.py

```diff
@@ -0,0 +1,125 @@
+
+    """
+    def __init__(self, container_client: ContainerClient):
+        """
+
+        :param container_client: The Azure Blob Storage Container client.
+        """
+        self._container_client = container_client
+        self._cached_ownership_dict = defaultdict(dict)  # type: Dict[str, Dict[str, Any]]
+        # lock each partition for list_ownership, claim_ownership and update_checkpoint etag doesn't get out of sync
+        # when the three methods are running concurrently
+        self._cached_ownership_locks = defaultdict(asyncio.Lock)
+
+    async def list_ownership(self, eventhub_name: str, consumer_group_name: str) -> Iterable[Dict[str, Any]]:
+        try:
+            blobs = self._container_client.list_blobs(include=['metadata'])
+        except AzureError as azure_err:  # list_blobs has exhausted retry
+            logger.exception("An exception occurred during list_ownership for eventhub %r consumer group %r."
+                             " An empty ownership list is returned",
+                             eventhub_name, consumer_group_name, exc_info=azure_err)
+            return []
+        async for b in blobs:  # TODO: running them concurrently
+            async with self._cached_ownership_locks[b.name]:
+                metadata = b.metadata
+                ownership = {
+                    "eventhub_name": eventhub_name,
+                    "consumer_group_name": consumer_group_name,
+                    "partition_id": b.name,
+                    "owner_id": metadata["owner_id"],
+                    "etag": b.etag,
+                    "last_modified_time": b.last_modified.timestamp() if b.last_modified else None
+                }
+                ownership.update(metadata)
+                self._cached_ownership_dict[b.name] = ownership
+        return self._cached_ownership_dict.values()
+
+    async def claim_ownership(self, ownership_list: Iterable[Dict[str, Any]]) -> Iterable[Dict[str, Any]]:
+        result = []
+        for ownership in ownership_list:  # TODO: claiming concurrently
+            partition_id = ownership["partition_id"]
+            eventhub_name = ownership["eventhub_name"]
+            consumer_group_name = ownership["consumer_group_name"]
+            owner_id = ownership["owner_id"]
+
+            async with self._cached_ownership_locks[partition_id]:
+                metadata = {"owner_id": ownership["owner_id"]}
+                if "offset" in ownership:
+                    metadata["offset"] = ownership["offset"]
+                if "sequence_number" in ownership:
+                    metadata["sequence_number"] = ownership["sequence_number"]
+
+                etag = ownership.get("etag")
+                if etag:
+                    etag_match = {"if_match": '"'+etag+'"'}
+                else:
+                    etag_match = {"if_none_match": '"*"'}
+                try:
+                    blob_client = await self._container_client.upload_blob(
```

Comment:
There is a race condition here. Please use the blob client instead of the container client.

Reply:
will change